websubhub-ballerina/hub_client.bal

```diff
@@ -0,0 +1,148 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/mime;
+import ballerina/crypto;
+
+# HTTP Based client for WebSub content publishing to subscribers
+public client class HubClient {
+    private string callback;
+    private string hub;
+    private string topic;
+    private string linkHeaderValue;
+    private string secret = "";
+    private http:Client httpClient;
+
+    # Initializes the `websubhub:HubClient`.
+    # ```ballerina
+    # websubhub:HubClient hubClientEP = new({
+    #   hub: "https://hub.com",
+    #   hubMode: "subscribe", 
+    #   hubCallback: "http://subscriber.com/callback", 
+    #   hubTopic: "https://topic.com", 
+    #   hubSecret: "key"
+    # });
+    # ```
+    #
+    # + url    - The URL to publish/notify updates
+    # + config - The `http:ClientConfiguration` for the underlying client or else `()`
+    public function init(Subscription subscription, http:ClientConfiguration? config = ()) returns error? {
+        self.callback = subscription.hubCallback;
+        self.hub = subscription.hub;
+        self.topic = subscription.hubTopic;
+        self.linkHeaderValue = generateLinkUrl(self.hub,  self.topic);
+        self.secret = subscription?.hubSecret is string ? <string>subscription?.hubSecret : "";
+        self.httpClient = check new(subscription.hubCallback, config);
+    }
+
+    # Distributes the published content to subscribers.
+    # ```ballerina
+    # error? publishUpdate = websubHubClientEP->notifyContentDistribution({
+    #   content: "This is sample content"
+    # });
+    #  ```
+    #
+    # + msg - content to be distributed to the topic-subscriber 
+    # + return -  An `error`if an error occurred with the update or else `()`
+    remote function notifyContentDistribution(ContentDistributionMessage msg) returns @tainted ContentDistributionSuccess | SubscriptionDeletedError | error? {
+        http:Client httpClient = self.httpClient;
+
+        http:Request request = new;
+        
+        string contentType = self.retrieveContentType(msg.contentType, msg.content);
+
+        check request.setContentType(contentType);
+        
+        if (msg?.headers is map<string|string[]>) {
+            var headers = <map<string|string[]>>msg?.headers;
+            foreach var [header, values] in headers.entries() {
+                if (values is string) {
+                    request.addHeader(header, values);
+                } else {
+                    string headerValue = ";".'join(...<string[]>values) + ";";
+                    request.addHeader(header, headerValue);
```

Comment:
add header actually appends the header values together, you can simply call addHeader() method inside a for loop